keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+1. Provide a CLI implementation, via kubectl, that interacts with the Application API.
+1. Provide installation status and garbage collection for applications.
+1. Provide a standard way for applications to surface a basic health check to the UIs.
+1. Provide an explicit mechanism for applications to declare dependencies on another application.
+1. Promote interoperability among ecosystem tools and UIs by creating a standard that tools MAY implement.
+
+### Non-Goals
+
+1. Create a standard that all tools MUST implement. 
+1. Provide a way for UIs to surface metrics from an application.
+
+## Proposal
+In order to address the issues discussed in the [previous section](#motivation) we will introduce the Application Kind 
+to the apps Group of the Kubernetes API, add the Application Controller to the Controller Manager, and modify 
+kubectl to make use the of the Application Kind.
+
+### User Stories
+
+#### Application Creation
+As an application maintainer or tool developer, I can create an application that is interoperable with other tools and 
+has first class support in Kuberentes UIs.
+
+#### Application View
+As an application user or UI designer, I have a standard way to view the components and status of Kubernetes 
+applications.
+
+#### Application Health
+As an application developer, I can surface a basic health check for UIs to consume, and, as a UI developer, I have a 
+standard way to consume application level health checks.
+
+#### Application Update
+As an application developer, maintainer, or administrator, I can update the configuration of my application.
+
+#### Application List
+As an application developer, maintainer or administrator, I can list the applications that are currently installed.
+
+#### Application Watch
+As an application developer, maintainer or administrator, I can watch the applications in a cluster for changes.
+
+#### Application Deletion
+As a user, I can delete my application, and when I delete my Application, its corresponding components are deleted.
+
+### Implementation
+This section describes the implementation of the feature, including the API, the associated controller, and the CLI 
+and UI interaction.
+
+### API
+The go code below describes the Application Kind. This code will be added to apps/v1alpha1 Group, and promoted through 
+apps/v1beta1 and apps/v1 Groups as stability of the Kind matures.
+
+```go
+
+// ApplicationStatus contains the status of an Application. 
+type ApplicationStatus struct {
+    // ObservedGeneration is used by the Application Controller to report the last Generation of an Application 
+    // that it has observed.
+    ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"int64,1,opt,name=observedGeneration"`
+    
+    // Installed is used by the Application Controller ot report the installed Components and Dependencies of  
+    // an Application.
+    Installed []string `json:"installed,omitempty" protobuf:"bytes,2,opt,name=installed"`
```

Comment:
It's not clear what goes in here as a string. It can't be `name` alone, because that underdetermines the reference. And there are ambiguities in the doc about dependencies also being tracked in this array.